drivers/usb/serial/xr_usb_serial_hal.c

```diff
@@ -0,0 +1,855 @@
+	{
+		xr_usb_serial_set_reg(xr_usb_serial,0xd02, preciseflags);
+	}
+	else if((xr_usb_serial->DeviceProduct == 0x1420)||
+		(xr_usb_serial->DeviceProduct == 0x1422)||
+		(xr_usb_serial->DeviceProduct == 0x1424))
+	{
+		xr_usb_serial_set_reg(xr_usb_serial, XR21B142X_WIDE_MODE_TX_OFFSET, preciseflags);
+		xr_usb_serial_set_reg(xr_usb_serial, XR21B142X_WIDE_MODE_RX_OFFSET, preciseflags);
+	}
+	xr_usb_serial_enable(xr_usb_serial);
+	return ret;
+}
+
+static int xr_usb_serial_tiocmget(struct xr_usb_serial *xr_usb_serial)
+{
+	short data;
+	int result;
+	result = xr_usb_serial_get_reg(xr_usb_serial,xr_usb_serial->reg_map.uart_gpio_status_addr, &data);
+	//dev_info(&xr_usb_serial->control->dev, "xr_usb_serial_tiocmget uart_gpio_status_addr:0x%04x\n",data);
+	if (result)
+		return ((data & 0x8) ? 0: TIOCM_DTR) | ((data & 0x20) ? 0:TIOCM_RTS ) | ((data & 0x4) ? 0:TIOCM_DSR) | ((data & 0x1) ? 0 : TIOCM_RI) | ((data & 0x2) ? 0:TIOCM_CD) | ((data & 0x10) ? 0 : TIOCM_CTS);
+	else
+		return -EFAULT;
+}
+
+static int xr_usb_serial_tiocmset(struct xr_usb_serial *xr_usb_serial,
+					unsigned int set, unsigned int clear)
+{
+	unsigned int newctrl = 0;
+	newctrl = xr_usb_serial->ctrlout;
+
+	set = (set & TIOCM_DTR ? XR_USB_SERIAL_CTRL_DTR : 0) | (set & TIOCM_RTS ? XR_USB_SERIAL_CTRL_RTS : 0);
+
+	clear = (clear & TIOCM_DTR ? XR_USB_SERIAL_CTRL_DTR : 0) | (clear & TIOCM_RTS ? XR_USB_SERIAL_CTRL_RTS : 0);
+
+	newctrl = (newctrl & ~clear) | set;
+
+	if (xr_usb_serial->ctrlout == newctrl)
+		return 0;
+
+	xr_usb_serial->ctrlout = newctrl;
+
+	if (newctrl & XR_USB_SERIAL_CTRL_DTR)
+		xr_usb_serial_set_reg(xr_usb_serial, xr_usb_serial->reg_map.uart_gpio_clr_addr, 0x08);
+	else
+		xr_usb_serial_set_reg(xr_usb_serial, xr_usb_serial->reg_map.uart_gpio_set_addr, 0x08);
+
+	if (newctrl & XR_USB_SERIAL_CTRL_RTS)
+		xr_usb_serial_set_reg(xr_usb_serial, xr_usb_serial->reg_map.uart_gpio_clr_addr, 0x20);
+	else
+		xr_usb_serial_set_reg(xr_usb_serial, xr_usb_serial->reg_map.uart_gpio_set_addr, 0x20);
+
+	return 0;
+}
+
+static struct reg_addr_map xr21b140x_reg_map;
+static struct reg_addr_map xr21b1411_reg_map;
+static struct reg_addr_map xr21v141x_reg_map;
+static struct reg_addr_map xr21b142x_reg_map;
+
+static void init_xr21b140x_reg_map(void)
+{
+	xr21b140x_reg_map.uart_enable_addr = 0x00;
+	xr21b140x_reg_map.uart_format_addr = 0x05;
+	xr21b140x_reg_map.uart_flow_addr = 0x06;
+	xr21b140x_reg_map.uart_loopback_addr = 0x16;
+	xr21b140x_reg_map.uart_xon_char_addr = 0x07;
+	xr21b140x_reg_map.uart_xoff_char_addr = 0x08;
+	xr21b140x_reg_map.uart_gpio_mode_addr = 0x0c;
+	xr21b140x_reg_map.uart_gpio_dir_addr = 0x0d;
+	xr21b140x_reg_map.uart_gpio_set_addr = 0x0e;
+	xr21b140x_reg_map.uart_gpio_clr_addr = 0x0f;
+	xr21b140x_reg_map.uart_gpio_status_addr = 0x10;
+	xr21b140x_reg_map.tx_break_addr = 0x0a;
+	xr21b140x_reg_map.uart_custom_driver = 0x41;
+}
+
+static void init_xr21b1411_reg_map(void)
+{
+	xr21b1411_reg_map.uart_enable_addr = 0xc00;
```

Comment:
xr21b1411 register map only differs from xr21b140x by a or with 0xc00, it would be more clear if there were defines for each register base...

Reply:
@rene again, I'm not modifying the original code, just adding rs485